btc_execution/sources/interpreter.move

```diff
@@ -1,6 +1,134 @@
 module btc_execution::interpreter;
+use btc_execution::opcode::isOpValid;
+use btc_execution::stack::{Stack, Self};
+
+
+// TODO: Follow error in btc implemetation
+#[error]
+const EBadReadData: vector<u8> = b"Invalid read script";
+#[error]
+const EBadOpcode: vector<u8> = b"Bad opcode";
+
+
+// Opcodes
+
+const OP_DUP: u8 = 76;
+
+
+public struct ScriptReader has copy, drop {
+    script: vector<u8>,
+    current_index: u64
+}
+
+public struct Interpreter has copy, drop {
+    stack: Stack
+}
+
+public fun reader(script: vector<u8>): ScriptReader {
+    ScriptReader {
+        script: script,
+        current_index: 0,
+    }
+}
+
+public fun create_interpreter(stack: Stack): Interpreter {
+    Interpreter {
+        stack : stack
+    }
+}
 
 /// Execute btc script
-public fun run(_script: vector<u8>) : bool {
-    return true
+public fun run(script: vector<u8>): bool {
+    let st = stack::create();
+    let mut ip = create_interpreter(st);
+    let mut r = reader(script);
+    ip.eval(&mut r)
+}
+
+fun eval(ip: &mut Interpreter, r: &mut ScriptReader): bool {
+
+    while(!r.end_stream()) {
+        let op = r.nextOpcode();
+
+        if (op == OP_DUP) {
+            ip.op_dup();
+            break
+        };
+    };
+
+    ip.isExecuteSuccess()
+}
+
+
+/// check evaluate is valid
+/// evaluation valid if the stack not empty
+/// and top element is non zero value
+public fun isExecuteSuccess(ip: &Interpreter): bool {
+    if (ip.stack.is_empty()) {
+        return false
+    };
+    let top = ip.stack.top();
+    cast_to_bool(&top)
```

Comment:
I think this should be the rule (please double check)
1. Stack length == 1 (no only non zero is checked)
2. the only item there is "true"

Reply:
BTC considers execution is success if: 
- The stack is not empty.
-  The top element is non-zero. 

Source from learnmeabitcoin: https://learnmeabitcoin.com/technical/script/#validity
In bitcoin-code: https://github.com/bitcoin/bitcoin/blob/master/src/script/interpreter.cpp#L1983

Reply:
The stack length can be greater than 1; the result depends on the top element.

Reply:
I think you are right:
> A transaction is valid if nothing in the combined script triggers failure and the top stack item is True (non-zero) when the script exits. The party that originally sent the Bitcoins now being spent dictates the script operations that will occur last in order to release them for use in another transaction. The party wanting to spend them must provide the input(s) to the previously recorded script that results in the combined script completing execution with a true value on the top of the stack. 

https://en.bitcoin.it/wiki/Script